src/utils/str.c

```diff
@@ -56,110 +56,198 @@ void str_printf(str *dst, const char *format, ...) {
-    dst->data[dst->len] = 0;
+void str_free(str *dst) {
+    if(dst == NULL) { return; }
+    omf_free(dst->data);
+    dst->len = 0;
 }
 
-void str_append(str *dst, const str *src) {
-    dst->data = omf_realloc(dst->data, dst->len + src->len + 1);
-    memcpy(dst->data + dst->len, src->data, src->len);
-    dst->len += src->len;
-    dst->data[dst->len] = 0;
+// ------------------------ Modification ------------------------
+
+
+void str_toupper(str *dst) {
+    for(size_t i = 0; i < dst->len; i++) {
+        dst->data[i] = toupper(dst->data[i]);
+    }
 }
 
-void str_append_c(str *dst, const char *src) {
-    size_t srclen = strlen(src);
-    dst->data = omf_realloc(dst->data, dst->len + srclen + 1);
-    memcpy(dst->data + dst->len, src, srclen);
-    dst->len += srclen;
-    dst->data[dst->len] = 0;
+void str_tolower(str *dst) {
+    for(size_t i = 0; i < dst->len; i++) {
+        dst->data[i] = tolower(dst->data[i]);
+    }
 }
 
-void str_prepend(str *dst, const str *src) {
-    dst->data = omf_realloc(dst->data, dst->len + src->len + 1);
-    memmove(dst->data + src->len, dst->data, dst->len);
-    memcpy(dst->data, src->data, src->len);
-    dst->len += src->len;
-    dst->data[dst->len] = 0;
+static size_t _strip_size(const str *src, bool left) {
+    if(src->len == 0) {
+        return 0;
+    }
+    size_t pos;
+    for(size_t i = 0; i < src->len; i++) {
+        pos = left ? i : src->len - i - 1;
+        if(!isspace(src->data[pos])) {
+            return pos;
+        }
+    }
+    return 0;
 }
 
-void str_prepend_c(str *dst, const char *src) {
-    size_t srclen = strlen(src);
-    dst->data = omf_realloc(dst->data, dst->len + srclen + 1);
-    memmove(dst->data + srclen, dst->data, dst->len);
-    memcpy(dst->data, src, srclen);
-    dst->len += srclen;
-    dst->data[dst->len] = 0;
+void str_rstrip(str *dst) {
+    // This is simple, just reduze sice and set ending 0.
+    size_t skip = _strip_size(dst, false);
+    STR_REALLOC(dst, skip + 1);
+    STR_ZERO(dst);
 }
 
-int str_first_of(const str *string, char find, size_t *pos) {
-    for(size_t i = 0; i < string->len; i++) {
+void str_lstrip(str *dst) {
+    // More complex. Move data first (memmmove!), then reduce size.
+    size_t skip = _strip_size(dst, true);
+    memmove(dst->data, dst->data + skip, dst->len - skip);
+    STR_REALLOC(dst, dst->len - skip);
+    STR_ZERO(dst);
+}
+
+void str_strip(str *dst) {
+    str_rstrip(dst);
+    str_lstrip(dst);
+}
+
+void str_append(str *dst, const str *src) {
+    assert(dst != src);
+    str_append_buf(dst, src->data, src->len);
+}
+
+void str_append_c(str *dst, const char *src) {
+    str_append_buf(dst, src, strlen(src));
+}
+
+void str_append_buf(str *dst, const char *buf, size_t len) {
+    size_t offset = dst->len;
+    STR_REALLOC(dst, dst->len + len);
+    memcpy(dst->data + offset, buf, len);
+    STR_ZERO(dst);
+}
+
+static bool _find_next(const str *string, char find, size_t *pos) {
```

Comment:
so i wonder why this underscore start on the static functions? They are static, no one is going to be using the functions outside this file anyway

Reply:
Kinda sorta reminder that "this function is static" without needing to check the declaration.

Reply:
Okay guess that's fine